[TOAZ-363] Allow UAMI accounts to be added to service admin whitelist

bennettn4 opened the pull request:
Ticket: (https://broadworkbench.atlassian.net/browse/TOAZ-363)

What:

UAMI accounts do not have OIDC_email field populated through accesstoken, need to be dynamically built from b2cid

Why:

Needed if Thurloe is running as a user assigned managed identity on azure and needs service account admin whitelist

How:

Additional check in service account admin whitelist based UAMI email

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://sdarq.dsp-appsec.broadinstitute.org/jira-ticket-risk-assesment) (requires Broad Internal network access) and attached the result to the JIRA ticket

bennettn4 committed:
allow uami accounts to be serviceAccountAdmins

UAMI accounts will not have oidc email field populated through token, need to be dynamically built
Needed if Thurloe is running as an UAMI and needs service account admin whitelist

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala:
```diff
@@ -49,7 +49,10 @@ trait StandardSamUserDirectives extends SamUserDirectives with LazyLogging with
     Directives
       .mapInnerRoute { r =>
-        if (!adminConfig.serviceAccountAdmins.contains(oidcHeaders.email)) {
+        if (
+          !adminConfig.serviceAccountAdmins.contains(oidcHeaders.email) &&
+          !adminConfig.serviceAccountAdmins.contains(WorkbenchEmail(s"${oidcHeaders.externalId.map(_.value).toOption.orNull}@uami.terra.bio"))
+        ) {
           reject(AuthorizationFailedRejection)
         } else {
           logger.info(s"Handling request for service admin account: ${oidcHeaders.email}")
```
I think this would be easier to follow in the affirmative, without the nots
```
        if (
          adminConfig.serviceAccountAdmins.contains(oidcHeaders.email) ||
          adminConfig.serviceAccountAdmins.contains(WorkbenchEmail(s"${oidcHeaders.externalId.map(_.value).toOption.orNull}@uami.terra.bio"))
        ) {
          // do the logic on starting on line 58
        } else {
          reject(AuthorizationFailedRejection)
```

bennettn4 replied:
Inverted boolean logic for readability in [ea35970](https://github.com/broadinstitute/sam/pull/1466/commits/ea359701ed0994c4ab517a75dafe378ab117b05c)

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardSamUserDirectives.scala:
```diff
@@ -49,7 +49,10 @@ trait StandardSamUserDirectives extends SamUserDirectives with LazyLogging with
-        if (!adminConfig.serviceAccountAdmins.contains(oidcHeaders.email)) {
+        if (
+          !adminConfig.serviceAccountAdmins.contains(oidcHeaders.email) &&
+          !adminConfig.serviceAccountAdmins.contains(WorkbenchEmail(s"${oidcHeaders.externalId.map(_.value).toOption.orNull}@uami.terra.bio"))
```
It bothers me that if the external id is absent, this ends up checking `null@uami.terra.bio`. I know that that is not real but it also should never be treated as valid. We also should only check for this kind of email if `oidcHeaders.email` is empty.
I suggest using a function like
```
  private def isAdminUami(oidcHeaders: OIDCHeaders) = {
    val maybeUamiEmail = oidcHeaders.externalId.map(b2cId => WorkbenchEmail(s"${b2cId.value}@uami.terra.bio")).toOption
    oidcHeaders.email.value.isEmpty && maybeUamiEmail.isDefined && adminConfig.serviceAccountAdmins.contains(maybeUamiEmail.get)
  }
```

bennettn4 replied:
Added suggested helper function in [ea35970](https://github.com/broadinstitute/sam/pull/1466/commits/ea359701ed0994c4ab517a75dafe378ab117b05c)

bennettn4 committed:
boolean logic update and helper function

Swapped boolean logic to allow if either condition is true as opposed to reject if both conditions are false
Added helper function to verify email oidcHeader is empty to clean up logic in route slightly and remove edge case of null@uami.terra.bio being a potentially valid value in serviceAccountAdmin configuration

bennettn4 committed:
Merge branch 'develop' into bennettn/sam-integration-thurloe-on-azure
